### Array Multiplication

Matrix multiplication is not regular multiplication:
- use the `@` operator like ---- A @ B
- use the `np.matmul(A, B)`

In [2]:
import numpy as np

In [3]:
A = np.random.randint(0,10,(2,2))
print(A)
B = np.random.randint(0,10,(2,2))
print(B)

A * B

[[0 7]
 [4 4]]
[[1 9]
 [0 1]]


array([[ 0, 63],
       [ 0,  4]])

In [4]:
A @ B

array([[ 0,  7],
       [ 4, 40]])

In [5]:
np.matmul(B, A)

array([[36, 43],
       [ 4,  4]])

### Aggregations
- Aggregations: - data and you summarize it
    - Sum
    - mean
    - min max
    - Standard Deviation
    - Variance
    - count
    - argmax argmin

- Aggregation along an AXIS
- `axis = 0` ===> down the ROWs --- (operate on Columns)
- `axis = 1` ===> across the COLUMNs --- (operate on Rows)

In [6]:
data = np.random.randint(0, 10, (2, 3))
print(data)

[[1 3 5]
 [1 5 2]]


In [7]:
print(data.sum()) # numpy array methods
print(sum(data)) # python way of data aggregation

print(data.mean())
print(data.std())
print(data.var())

print(data.min())
print(data.max())

17
[2 8 7]
2.8333333333333335
1.674979270186815
2.805555555555555
1
5


In [8]:
print(data.sum(axis= 0))
print(data.sum(axis= 1))

[2 8 7]
[9 8]


In [9]:
print(data.mean(axis= 0))
print(data.mean(axis= 1))

[1.  4.  3.5]
[3.         2.66666667]


### Boolean Masking - (Filtering)
- used to filter elements out
- 
- array[mask] ==== mask is a condition
- `array_name[<condition>]`

In [10]:
scores = np.random.randint(0, 101, 6)
scores

array([56,  7,  9, 35, 83, 77])

In [11]:
print(scores > 39)
print(scores[scores > 39])

[ True False False False  True  True]
[56 83 77]


In [12]:
age = np.array([5, 17, 24, 51, 23, 14, 8, 29, 33, 44, 10])
vote_age = age >= 18
age[vote_age]

array([24, 51, 23, 29, 33, 44])

In [13]:
voting_eligibility = np.where(age >= 18, 'Eligible', 'Not-Eligible')

In [14]:
list_zip = list(zip(age, voting_eligibility))

In [15]:
for age, voting_eligibility in list_zip:
    print(f'{age}: {voting_eligibility}!')

5: Not-Eligible!
17: Not-Eligible!
24: Eligible!
51: Eligible!
23: Eligible!
14: Not-Eligible!
8: Not-Eligible!
29: Eligible!
33: Eligible!
44: Eligible!
10: Not-Eligible!


In [16]:
age = np.array([5, 17, 24, 51, 23, 14, 8, 29, 33, 44, 10])
voting_eligibility = np.where(age >= 18, 'Eligible', 'Not-Eligible')

combined_array = np.column_stack((age, voting_eligibility))
print(combined_array)

[['5' 'Not-Eligible']
 ['17' 'Not-Eligible']
 ['24' 'Eligible']
 ['51' 'Eligible']
 ['23' 'Eligible']
 ['14' 'Not-Eligible']
 ['8' 'Not-Eligible']
 ['29' 'Eligible']
 ['33' 'Eligible']
 ['44' 'Eligible']
 ['10' 'Not-Eligible']]


---
BONUS PART

### Nested Loops

```python

for <variable> in <iterable>:
    print(<variables>)

In [17]:
for i in range(5):
    print(i)

0
1
2
3
4


In [18]:
alist = [3, 5, 7, 9]
name = 'Alya'

for variable in name:
    print(variable)

A
l
y
a


- enumerate object is a sequence
- just like zip

- just like range - range can be type casted into a list
    - others cannot

- we run for loops on them - 

In [19]:
for i, j in enumerate(name):
    print(i, j)

0 A
1 l
2 y
3 a


In [20]:
for i in range(3):
    for j in range(3):
        print(i)

0
0
0
1
1
1
2
2
2


In [21]:
array2d = np.random.randint(0, 10, (2, 4))
array2d

array([[5, 2, 4, 4],
       [6, 0, 8, 6]])

In [22]:
for row in array2d:
    for value in row:
        print(value)

5
2
4
4
6
0
8
6


In [23]:
for row in array2d:
    for value in row:
        print(value, end='  ')
    print()

5  2  4  4  
6  0  8  6  


In [24]:
for i in range(len(array2d)):
    for j in range(len(array2d[i])):
        print(f'Row {i}, col {j}: {array2d[i,j]}')

Row 0, col 0: 5
Row 0, col 1: 2
Row 0, col 2: 4
Row 0, col 3: 4
Row 1, col 0: 6
Row 1, col 1: 0
Row 1, col 2: 8
Row 1, col 3: 6


### Exercises

The Scenario:

- The school's database has just exported a batch of raw exam scores. You need to use NumPy to generate some sample data, inspect its structure, and automatically grade the students based on a curve.

Your Tasks:

1. The Setup: Import NumPy. To ensure the whole class gets the exact same "random" numbers, set the random seed to 42.

2. Data Generation: Use random.randint to generate a 2D array of random scores between 0 and 100. The array should represent 10 students and their scores across 3 different subjects.

3. Array Inspection: Prove you built the array correctly by printing its ndim (number of dimensions), its shape, and its total size (total number of scores).

4. Student IDs: Use arange to create a 1D array of Student IDs starting at 100 and ending at 110.

5. Perform the aggregate operation of average and find the average score of students for each subject

6. The Audit (Boolean Masking): The principal wants to know every instance where a student scored below a 50. Use boolean masking to extract and print just those failing scores from the main array.

7. Automated Grading (where): Use np.where to create a brand new "Report Card" array. If a score in the original array is less than 50, it should say "Fail". If it is 50 or above, it should say "Pass".

In [6]:
# Task 1: The Setup
import numpy as np
np.random.seed(42)

# Task 2: Data Generation
students_scores = np.random.randint(0, 101, (10, 3))
print(students_scores)

# Task 3: Array Inspection
print(students_scores.ndim)
print(students_scores.shape)
print(students_scores.size)

# Task 4: Student IDs
student_IDs = np.arange(100, 110)
print(student_IDs)

# Task 5: aggregate
print(students_scores.mean(axis= 0))

# Task 6: The Audit
print(students_scores[students_scores < 50])

# Task 7: Automated Grading
report_card = np.where(students_scores < 50, 'Fail', 'Pass')
print(report_card)

[[51 92 14]
 [71 60 20]
 [82 86 74]
 [74 87 99]
 [23  2 21]
 [52  1 87]
 [29 37  1]
 [63 59 20]
 [32 75 57]
 [21 88 48]]
2
(10, 3)
30
[100 101 102 103 104 105 106 107 108 109]
[49.8 58.7 44.1]
[14 20 23  2 21  1 29 37  1 20 32 21 48]
[['Pass' 'Pass' 'Fail']
 ['Pass' 'Pass' 'Fail']
 ['Pass' 'Pass' 'Pass']
 ['Pass' 'Pass' 'Pass']
 ['Fail' 'Fail' 'Fail']
 ['Pass' 'Fail' 'Pass']
 ['Fail' 'Fail' 'Fail']
 ['Pass' 'Pass' 'Fail']
 ['Fail' 'Pass' 'Pass']
 ['Fail' 'Pass' 'Fail']]


The Scenario 2:

The semester is over. The administration has updated the grading policy: if a student fails even one subject (scores below 50), their overall status becomes "Fail". The principal needs you to generate a final report that includes Student IDs, their scores, and their final Pass/Fail status. You must combine this data into one final grid, export it, and try to apply a last-minute bonus.

Your Tasks:

(Assume you already have your results array of 10 students across 3 subjects, and your student_ID array from 100 to 109).

1. Column Extraction (2D Slicing): Using 2D slicing, extract each subject from the results array into its own 1D variable. You should end up with `subject_1`, `subject_2`, and `subject_3`.

2. The Strict Grading Logic (`np.where`): Use `np.where` to create a `strict_status` array. If a student scored below 50 in Subject 1, OR Subject 2, OR Subject 3, their status is "Fail". Otherwise, it is "Pass".

    - Hint: In standard Python you use the word or, but in NumPy, you must use the Bitwise OR operator (|).

    - Crucial Rule: When combining multiple conditions in NumPy, you must wrap each individual condition in parentheses! `(condition 1) | (condition 2)`

3. The Big Merge (column_stack): The principal wants all data side-by-side. Use `np.column_stack` to merge the `student_ID` array, the original `2D results array`, and your new `strict_status` array into one single 2D array called `final_report`. Print it to the screen.

4. The NumPy Trap (Data Types): The administration suddenly wants you to add 5 bonus points to the scores in your final_report. Try to add 5 to one of the score columns. What happens? Print the .dtype (data type) of your final_report to figure out why.

5. The Export: Even though the math broke, save your final_report to your computer as a CSV file called "final_grades.csv" so another program can deal with it.

> Hint: np.savetxt("document_name.csv", , delimiter=",", fmt="%s", header="ID,Sub1,Sub2,Sub3,Status")